trimesh/triangles.py

```diff
@@ -253,32 +253,26 @@ def bounds_tree(triangles):
 
 def nondegenerate(triangles, areas=None):
     """
-    Find all triangles which have an oriented bounding box
-    where both of the two sides is larger than tol.merge and
-    both edge vectors are longer than tol.merge.
-
-    Degenerate triangles can be when:
-    1) Two of the three vertices are colocated
-    2) All three vertices are unique but colinear
-
+    Return a mask that identifies triangles without three collinear vertices.
 
     Parameters
     ----------
     triangles: (n, 3, 3) float, list of triangles
 
-
     Returns
     ----------
-    nondegenerate: (n,) bool array of triangles that have area
+    nondegenerate: (n,) bool array where True indicates a nondegenerate tri.
     """
     triangles = np.asanyarray(triangles, dtype=np.float64)
     if not util.is_shape(triangles, (-1, 3, 3)):
         raise ValueError('Triangles must be (n,3,3)!')
 
-    # if both edges of the triangles OBB are longer than tol.merge
-    # we declare them to be nondegenerate
-    ok = (extents(triangles=triangles,
-                  areas=areas) > tol.merge).all(axis=1)
+    # Minimum angle threshold
+    a, valid = util.unitize(triangles[:, 1] - triangles[:, 0], check_valid=True)
+    b, valid = util.unitize(triangles[:, 2] - triangles[:, 1], check_valid=True)
+    dots = np.abs(np.einsum('ij,ij->i', a[valid], b[valid]))
+    ok = valid
+    ok[valid] = (1.0 - dots > tol.zero)
```

Comment:
New change here for you to review -- basically does a double filter for truly zero-area faces and faces whose points are collinear by some threshold.

Reply:
Could also re-do this to just explictly compute normals and then mark any faces with zero-area for deletion.

Reply:
Gotcha, so this formulation of degenerate faces is saying, "both edges are longer than the tolerance inside unitize, and the dot product of one onto the other is larger than tol.zero." Seems reasonable- also I changed the threshold to be settable in unitize in a branch. 

Though doesn't this work out to the same thing as the original one which was "one edge and the height of the triangle are greater than tol.merge"? The dot product is the height, and the edge length is the width. BTW in the formulation above it should probably be
```
a, a_valid = util.unitize(triangles[:, 1] - triangles[:, 0], check_valid=True)
b, b_valid = util.unitize(triangles[:, 2] - triangles[:, 1], check_valid=True)
valid = np.logical_and(a_valid, b_valid)
```

Your point about preferring merged vertices over degenerate faces makes sense- so we should only be throwing away faces significantly smaller than tol.merge (tol_zero seems good).

Reply:
Oh nice einsum btw, I benchmarked to see if I should switch `trimesh.util.diagonal_dot` to use it, looks like the sum method is slightly faster albeit with more variance for some reason: 

```
In [11]: %timeit trimesh.util.diagonal_dot(a,b)
The slowest run took 14.29 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 3: 22.3 µs per loop

In [12]: %timeit np.einsum('ij,ij->i', a,b)
The slowest run took 5.41 times longer than the fastest. This could mean that an intermediate result is being cached.
10000 loops, best of 3: 51.8 µs per loop


```

Reply:
Good catch on the formulation bug, I'll fix it and update :)

I don’t think this is the same thing as the original one. Here’s my thinking:

There are two types of degenerate faces:
- Faces with co-located vertices (an edge less than `tol.merge` in length)
- Faces with unique vertices, but the vertices are collinear

The first case is taken care of by `merge_vertices`, so any triangle passed to this function is guaranteed to have three edges with length at least `tol.merge`.

The second case should be addressed by this function. In the original implementation, the check was dependent on the area of the triangle, but when checking for collinearity, we probably don’t want to give different results if the mesh is simply scaled up. In the new implementation, I explicitly unitize the vectors to factor out scale, and then threshold the angle between them (the dot product). I think that this is probably cleaner, as it creates a clearer criterion for rejecting faces. What do you think?

An example of where the new function would differ from the old is with an equilateral triangle with side length `1.01e-8`. It seems that this should be allowed, but the old method would reject it, as the height of this triangle is less than `1e-8`. However, the new method would note that the angles are fine (60 degrees) and not flag the triangle as bad.

Actually, I probably don’t even need the validity check for `unitize` here — all of the edges must be greater than `1e-8` in length, since they were already fixed up by merging vertices. I’m thinking it can’t hurt to keep it, though :)

Let me know what you think! Sorry this is turning into a long PR comment chain haha, these sort of thresholding things tend to be messy. Definitely don’t change anything just for my use case!!! Only accept them if you think they make good sense in the general case.

Best,
Matt

Reply:
Actually, my method does have a failure case -- very long, skinny triangles :( I think your method might actually be better with some re-jiggering... going to think about this tonight while I sleep haha.